File: 0206-reverse-linked-list/0206-reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
//   init previous variable
  var prevNode = null;
//   init current node variable for traversal
  var currNode = head;
  
//   iterate while current node isn't empty
  while (currNode !== null) {
//     store next value in a temp variable so we don't lose it
    var tempNext = currNode.next;
//     reverse the connection
    currNode.next = prevNode;
//     update prev to current
    prevNode = currNode;
//     update curr to true next
    currNode = tempNext;
  }
  
//   when you reach the end of the list, return the head (which is prev at this point)
  return prevNode;
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # sentinel node/dummy node becomes the new tail
        # floating next reference to the next node
        # traverse initial order while simultaneously reversing the links

        # need a prev node
        # iterate through current list starting from the head (iteration stops when curr node is null)
          # store reference to next node
          # set current node's next to the node behind it
          # current node becomes next node that we stored a reference to earlier

        prev = None
        curr = head
        while curr:
          next = curr.next
          curr.next = prev
          prev = curr
          curr = next
        
        return prev

        

